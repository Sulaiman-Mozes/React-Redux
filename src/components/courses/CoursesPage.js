import React from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import { Redirect } from 'react-router-dom';
import { toast } from 'react-toastify';
import { loadCourses, deleteCourse } from '../../redux/actions/CourseActions';
import { loadAuthors } from '../../redux/actions/AuthorActions';
import CourseList from './CourseList';
import Spinner from '../commons/Spinner';

class CoursesPage extends React.PureComponent {
  state = {
    redirectToAddCoursePage: false,
  }

  componentDidMount() {
    const { getCourses, getAuthors } = this.props;
    getCourses();
    getAuthors();
  }

  handelDelete = async (course) => {
    toast.success('Course Deleted');
    const { removeCourse } = this.props;
    try {
      await removeCourse(course);
    } catch (err) {
      toast.error(`Delete Failed. ${err.message}`, { autoClose: false });
    }
  }

  render() {
    const { courses, authorLoading, courseLoading } = this.props;
    const { redirectToAddCoursePage } = this.state;
    return (
      <>
        {redirectToAddCoursePage && <Redirect to="/course" />}
        <p>Courses</p>
        <button
          type="button"
          style={{ marginBottom: 20 }}
          className="btn btn-primary add-course"
          onClick={() => this.setState({ redirectToAddCoursePage: true })}
        >
          Add Course
        </button>
        {authorLoading || courseLoading
          ? (<Spinner />)
          : (
            <CourseList
              courses={courses}
              onDelete={this.handelDelete}
            />
          )
        }
      </>
    );
  }
}

CoursesPage.propTypes = {
  getCourses: PropTypes.func.isRequired,
  getAuthors: PropTypes.func.isRequired,
  removeCourse: PropTypes.func.isRequired,
  authorLoading: PropTypes.bool.isRequired,
  courseLoading: PropTypes.bool.isRequired,
  courses: PropTypes.arrayOf(PropTypes.object).isRequired,
};

const mapStateToProps = ({ courses, authors }) => ({
  courses: authors.list.length === 0 ? [] : courses.list.map(course => ({
    ...course,
    authorName: authors.list.find(a => a.id === course.authorId).name,
  })),
  authorLoading: authors.loading,
  courseLoading: courses.loading,
});

const mapDispatchToProps = {
  getCourses: loadCourses,
  getAuthors: loadAuthors,
  removeCourse: deleteCourse,
};

export default connect(mapStateToProps, mapDispatchToProps)(CoursesPage);
