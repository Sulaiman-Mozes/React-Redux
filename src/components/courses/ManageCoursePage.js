import React, { useEffect, useState } from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import { toast } from 'react-toastify';
import { saveCourse, loadCourses } from '../../redux/actions/CourseActions';
import { loadAuthors } from '../../redux/actions/AuthorActions';
import CourseForm from './CourseForm';
import { newCourse } from '../../../tools/mockData';
import Spinner from '../commons/Spinner';

export const ManageCoursesPage = ({
  getCourses, getAuthors, history, courseLoading, authorLoading,
  addCourse, courses, authors, ...props
}) => {
  const [course, setCourse] = useState({ ...props.course });
  const [errors, setErrors] = useState({});
  const [saving, setSaving] = useState(false);

  useEffect(() => {
    getCourses();
    getAuthors();
    setCourse({ ...props.course });
  }, []);

  const handleChange = (event) => {
    const { name, value } = event.target;
    setCourse(previousState => ({
      ...previousState,
      [name]: name === 'authorId' ? parseInt(value, 10) : value,
    }));
  };

  const formIsValid = () => {
    const { title, authorId, category } = course;
    const error = {};
    if (!title) error.title = 'Title is required.';
    if (!authorId) error.author = 'Author is required.';
    if (!category) error.category = 'Category is required.';

    setErrors(error);
    return Object.keys(errors).length === 0;
  };

  const handleSubmit = (event) => {
    event.preventDefault();
    if (!formIsValid()) return;
    setSaving(true);
    addCourse(course).then(() => {
      toast.success('Course Successfully Added');
      history.push('/courses');
    }).catch((error) => {
      setSaving(false);
      setErrors({ onSave: error.message });
    });
  };

  return (
    <>
      {authorLoading || courseLoading
        ? (<Spinner />)
        : (
          <CourseForm
            course={course}
            errors={errors}
            authors={authors}
            onChange={handleChange}
            onSave={handleSubmit}
            saving={saving}
          />
        )
      }
    </>
  );
};

ManageCoursesPage.propTypes = {
  addCourse: PropTypes.func.isRequired,
  getCourses: PropTypes.func.isRequired,
  getAuthors: PropTypes.func.isRequired,
  course: PropTypes.shape({}).isRequired,
  authorLoading: PropTypes.bool.isRequired,
  courseLoading: PropTypes.bool.isRequired,
  history: PropTypes.shape({}).isRequired,
  courses: PropTypes.arrayOf(PropTypes.object).isRequired,
  authors: PropTypes.arrayOf(PropTypes.object).isRequired,
};

export const getCourseBySlug = (courses, slug) => courses.find(
  course => course.slug === slug,
) || newCourse;

const mapStateToProps = ({ courses, authors }, ownProps) => {
  const { params: { slug } } = ownProps.match;
  const course = slug && courses.list.length > 0
    ? (getCourseBySlug(courses.list, slug)) : newCourse;
  return {
    course,
    courses: courses.list,
    authors: authors.list,
    authorLoading: authors.loading,
    courseLoading: courses.loading,
  };
};

const mapDispatchToProps = {
  addCourse: saveCourse,
  getCourses: loadCourses,
  getAuthors: loadAuthors,
};

export default connect(mapStateToProps, mapDispatchToProps)(ManageCoursesPage);
