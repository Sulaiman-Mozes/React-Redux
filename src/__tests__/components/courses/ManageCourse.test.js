import React from 'react';
import { mount } from 'enzyme';
import { ManageCoursesPage } from '../../../components/courses/ManageCoursePage';
import { authors, courses } from '../../../../tools/mockData';


describe('Test Form Component', () => {
  const render = (args) => {
    const defaultProps = {
      courses: authors,
      authors: courses,
      history: {},
      addCourse: () => ({
        then: () => ({ catch: () => { } }),
      }),
      getCourses: jest.fn(),
      getAuthors: jest.fn(),
      course: {
        id: null,
        title: '',
        authorId: null,
        category: '',
      },
      authorLoading: false,
      courseLoading: false,
    };

    const props = { ...defaultProps, ...args };
    return mount(<ManageCoursesPage {...props} />);
  };

  it('Should Render Correctly', () => {
    expect(render()).toMatchSnapshot();
  });

  it('Sets an error on atttempting to submit an empty from', () => {
    const wrapper = render();
    wrapper.find('form').simulate('submit');
    const error = wrapper.find('.alert').first();
    expect(error.text()).toBe('Title is required.');
  });

  it('Should Render Spinner When Loading', () => {
    const args = {
      courses: [],
      authors: [],
      authorLoading: true,
      courseLoading: true,
    };
    expect(render(args).find('Spinner')).toHaveLength(1);
  });
});
