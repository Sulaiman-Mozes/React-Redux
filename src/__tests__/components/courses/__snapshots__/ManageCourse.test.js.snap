// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Form Component Should Render Correctly 1`] = `ReactWrapper {}`;
